build fails on nixos 15.09

I'm working on a [package](https://gist.github.com/7f874e1e0b04f2e42e46) for [NixOS](http://nixos.org).

All versions that I've tried give me:

```
-- Boost version: 1.58.0
  CMake Error at build/CMakeFiles/git-data/grabRef.cmake:36 (file):
    file failed to open for reading (No such file or directory):
      /tmp/nix-build-i-score-v1.0.0-a31.drv-0/i-score-6037d70/build/CMakeFiles/git-data/head-ref
  Call Stack (most recent call first):
  CMake/modules/GetGitRevisionDescription.cmake:77 (include)
  CMake/IScoreConfiguration.cmake:198 (get_git_head_revision)
  CMakeLists.txt:14 (include)
-- cotire 1.7.5 loaded.
```

When I `sed -e '77d' -i CMake/modules/GetGitRevisionDescription.cmake` the error is gone.

From v1.0.0-a37 or v1.0.0-a38 on (haven't tried yet), I get:

```
 -- The CXX compiler identification is unknown
 CMake Error at CMakeLists.txt:14 (project):
   The CMAKE_CXX_COMPILER:
       /var/empty/bin/clang++
         is not a full path to an existing compiler tool.
```

Version v1.0.0-a36 and older don't do this, but give me other problems.

@jcelerier Thanks!
I've updated [the package](https://gist.github.com/magnetophon/3c9271e41947ea803932) according to your suggestions.
Unfortunately, the build still fails:

```
 [709/777] Building CXX object base/plugins/iscore-plugin-js/CMakeFiles/iscore_plugin_js.dir/JS/JSProcessModelSerialization.cpp.o
FAILED: /nix/store/9fbfiij3ajnd3fs1zyc2qy0ispbszrr7-gcc-wrapper-4.9.3/bin/g++   -DISCORE_LIB_BASE -DISCORE_LIB_DEVICE -DISCORE_LIB_DUMMYPROCESS -DISCORE_LIB_INSPECTOR -DISCORE_LIB_PROCESS -DISCORE_LIB_STATE -DISCORE_PLUGIN_AUTOMATION -DISCORE_PLUGIN_CURVE -DISCORE_PLUGIN_DEVICEEXPLORER -DISCORE_PLUGIN_LOOP -DISCORE_PLUGIN_OSSIA -DISCORE_PLUGIN_SCENARIO -DQT_CORE_LIB -DQT_GUI_LIB -DQT_NO_DEBUG -DQT_SVG_LIB -DQT_WIDGETS_LIB -DQT_XML_LIB -Discore_plugin_space_EXPORTS -O3 -DNDEBUG -fPIC -fvisibility=hidden -fvisibility-inlines-hidden -Ibase/plugins/iscore-plugin-space -I../base/plugins/iscore-plugin-space -I/nix/store/lfj41wlg1m05rfip5jja4mpfp84yiyhm-boost-1.58.0-dev/include -I. -isystem ../3rdparty/variant/include -isystem ../3rdparty/nano-signal-slot -I../base/lib -I../ -isystem /nix/store/nc1fslq47z4qpizwdjx97fjj6x7bvnd4-qtbase-5.4.2/include -isystem /nix/store/nc1fslq47z4qpizwdjx97fjj6x7bvnd4-qtbase-5.4.2/include/QtCore -isystem /nix/store/nc1fslq47z4qpizwdjx97fjj6x7bvnd4-qtbase-5.4.2/mkspecs/linux-g++ -isystem /nix/store/nc1fslq47z4qpizwdjx97fjj6x7bvnd4-qtbase-5.4.2/include/QtWidgets -isystem /nix/store/nc1fslq47z4qpizwdjx97fjj6x7bvnd4-qtbase-5.4.2/include/QtGui -isystem /nix/store/rj0iyafchrdppzyidn5xz8q5bvnk1iv1-mesa-noglu-10.6.9/include -Ibase/lib -isystem /nix/store/abm9qnv54s9s9sb6sl3mdqvliwkhqf0c-qtsvg-5.4.2/include -isystem /nix/store/abm9qnv54s9s9sb6sl3mdqvliwkhqf0c-qtsvg-5.4.2/include/QtSvg -I../3rdparty/QRecentFilesMenu -Ibase/plugins/iscore-plugin-deviceexplorer -I../base/plugins/iscore-plugin-deviceexplorer -Ibase/plugins/iscore-lib-state -I../base/plugins/iscore-lib-state -Ibase/plugins/iscore-lib-device -I../base/plugins/iscore-lib-device -isystem /nix/store/nc1fslq47z4qpizwdjx97fjj6x7bvnd4-qtbase-5.4.2/include/QtXml -I../3rdparty/QProgressIndicator -Ibase/plugins/iscore-lib-process -I../base/plugins/iscore-lib-process -Ibase/plugins/iscore-plugin-ossia -I../base/plugins/iscore-plugin-ossia -Ibase/plugins/iscore-plugin-scenario -I../base/plugins/iscore-plugin-scenario -Ibase/plugins/iscore-plugin-curve -I../base/plugins/iscore-plugin-curve -Ibase/plugins/iscore-plugin-inspector -I../base/plugins/iscore-plugin-inspector -Ibase/plugins/iscore-plugin-automation -I../base/plugins/iscore-plugin-automation -I../API/Headers -isystem /nix/store/f8a98daj1gnn1jf22grz3dvfa8dgjrcr-JamomaCore-v1.0-beta.1/jamoma/include/jamoma -Ibase/plugins/iscore-plugin-loop -I../base/plugins/iscore-plugin-loop -Ibase/plugins/iscore-plugin-scenario/DummyProcess -I../base/plugins/iscore-plugin-scenario/DummyProcess    -Wall -Wextra -Wno-unused-parameter -Wno-unknown-pragmas -Wnon-virtual-dtor -pedantic -Woverloaded-virtual -pipe -Ofast -fPIC -std=gnu++14 -MMD -MT base/plugins/iscore-plugin-space/CMakeFiles/iscore_plugin_space.dir/src/SpaceProcessFactory.cpp.o -MF base/plugins/iscore-plugin-space/CMakeFiles/iscore_plugin_space.dir/src/SpaceProcessFactory.cpp.o.d -o base/plugins/iscore-plugin-space/CMakeFiles/iscore_plugin_space.dir/src/SpaceProcessFactory.cpp.o -c ../base/plugins/iscore-plugin-space/src/SpaceProcessFactory.cpp
In file included from ../base/plugins/iscore-plugin-space/Space/computation.hpp:4:0,
                 from ../base/plugins/iscore-plugin-space/src/Computation/ComputationModel.hpp:3,
                 from ../base/plugins/iscore-plugin-space/src/SpaceProcess.hpp:5,
                 from ../base/plugins/iscore-plugin-space/src/SpaceProcessFactory.cpp:2:
../base/plugins/iscore-plugin-space/Space/square_approximator.hpp: In member function 'constexpr spacelib::square_approx<size, side>::iterator spacelib::square_approx<size, side>::iterator::operator++() const':
../base/plugins/iscore-plugin-space/Space/square_approximator.hpp:27:28: error: increment of read-only location '((const spacelib::square_approx<size, side>::iterator*)this)->spacelib::square_approx<size, side>::iterator::pos'
                 { return ++pos, *this; }
                            ^
../base/plugins/iscore-plugin-space/Space/square_approximator.hpp: In member function 'constexpr spacelib::square_approx<size, side>::iterator spacelib::square_approx<size, side>::iterator::operator++(int) const':
../base/plugins/iscore-plugin-space/Space/square_approximator.hpp:29:28: error: increment of read-only location '((const spacelib::square_approx<size, side>::iterator*)this)->spacelib::square_approx<size, side>::iterator::pos'
                 { return ++pos, *this; }
                            ^
../base/plugins/iscore-plugin-space/Space/square_approximator.hpp: In member function 'constexpr spacelib::dynamic_square_approx::iterator spacelib::dynamic_square_approx::iterator::operator++() const':
../base/plugins/iscore-plugin-space/Space/square_approximator.hpp:74:28: error: increment of member 'spacelib::dynamic_square_approx::iterator::m_pos' in read-only object
                 { return ++m_pos, *this; }
```

--- a/src/plugins/score-plugin-vst3/Vst3/EffectModel.hpp
+++ b/src/plugins/score-plugin-vst3/Vst3/EffectModel.hpp
@@ -6,6 +6,7 @@
 #include <Control/DefaultEffectItem.hpp>
 #include <Effect/EffectFactory.hpp>
 #include <Vst3/Plugin.hpp>
+#include <Vst3/UI/WindowContainer.hpp>
 
 #include <score/tools/std/Invoke.hpp>
 
@@ -36,7 +37,8 @@ namespace vst3
 class CreateVSTControl;
 class ControlInlet;
 struct PortCreationVisitor;
-
+class PlugFrame;
+class Window;
 class Model final : public Process::ProcessModel
 {
   W_OBJECT(Model)

--- a/src/plugins/score-plugin-vst3/Vst3/Plugin.cpp
+++ b/src/plugins/score-plugin-vst3/Vst3/Plugin.cpp
@@ -2,6 +2,8 @@
 #include <Vst3/DataStream.hpp>
 #include <Vst3/EditHandler.hpp>
 #include <Vst3/Plugin.hpp>
+#include <Vst3/UI/Linux/PlugFrame.hpp>
+#include <Vst3/UI/PlugFrame.hpp>
 #include <Vst3/UI/Window.hpp>
 
 #include <ossia/detail/algorithms.hpp>
@@ -240,6 +242,12 @@ void Plugin::start(double_t sample_rate, int max_bs)
 
 void Plugin::stop()
 {
+#if (!(defined(__APPLE__) || defined(_WIN32))) && __has_include(<xcb/xcb.h>)
+  if(plugFrame)
+  {
+    plugFrame->cleanup();
+  }
+#endif
   if(!component)
     return;
 
@@ -263,6 +271,11 @@ void Plugin::stop()
 
   component->terminate();
   component = nullptr;
+
+  if(plugFrame)
+  {
+    QTimer::singleShot(100, [=] { delete plugFrame; });
+  }
 }
 
 Plugin::~Plugin() { }

--- a/src/plugins/score-plugin-vst3/Vst3/Plugin.hpp
+++ b/src/plugins/score-plugin-vst3/Vst3/Plugin.hpp
@@ -38,7 +38,7 @@ inline QString fromString(const Steinberg::Vst::String128& str)
 }
 
 using MIDIControls = ossia::flat_map<std::pair<int, int>, Steinberg::Vst::ParamID>;
-
+class PlugFrame;
 struct Plugin
 {
   Plugin() = default;
@@ -61,6 +61,7 @@ struct Plugin
   Steinberg::Vst::IEditController* controller{};
   Steinberg::Vst::IUnitInfo* units{};
   Steinberg::IPlugView* view{};
+  PlugFrame* plugFrame{};
 
   void loadAudioProcessor(ApplicationPlugin& ctx);
   void loadEditController(Model& model, ApplicationPlugin& ctx);

--- a/src/plugins/score-plugin-vst3/Vst3/UI/Linux/PlugFrame.hpp
+++ b/src/plugins/score-plugin-vst3/Vst3/UI/Linux/PlugFrame.hpp
@@ -201,17 +201,20 @@ public:
 
     auto readnotifier = new QSocketNotifier{fd, QSocketNotifier::Read};
     readnotifier->setEnabled(true);
+    readnotifier->setParent(&internalContextObject);
     auto on_fd = [=] { handler->onFDIsSet(fd); };
     QObject::connect(
         readnotifier, &QSocketNotifier::activated, &internalContextObject, on_fd);
 
     auto writenotifier = new QSocketNotifier{fd, QSocketNotifier::Write};
     writenotifier->setEnabled(true);
+    writenotifier->setParent(&internalContextObject);
     QObject::connect(
         writenotifier, &QSocketNotifier::activated, &internalContextObject, on_fd);
 
     auto errnotifier = new QSocketNotifier{fd, QSocketNotifier::Exception};
     errnotifier->setEnabled(true);
+    errnotifier->setParent(&internalContextObject);
     QObject::connect(
         errnotifier, &QSocketNotifier::activated, &internalContextObject, on_fd);
 
@@ -255,6 +258,7 @@ public:
       Linux::ITimerHandler* handler, Linux::TimerInterval milliseconds) override
   {
     auto t = new QTimer;
+    t->setParent(&internalContextObject);
     QObject::connect(t, &QTimer::timeout, [=] { handler->onTimer(); });
     t->start(milliseconds);
     timers.push_back({handler, t});
@@ -272,18 +276,19 @@ public:
     return Steinberg::kResultOk;
   }
 
-  QDialog& w;
+  QDialog* w;
   WindowContainer wc;
   explicit PlugFrame(QDialog& w, WindowContainer wc)
-      : w{w}
+      : w{&w}
       , wc{wc}
   {
   }
 
-  ~PlugFrame()
+  void cleanup()
   {
     for(auto timer : timers)
       delete timer.second;
+    timers.clear();
     for(auto handler : handlers)
     {
       GlobalSocketHandlers::instance().unregisterHandler(handler.handler, handler.fd);
@@ -291,11 +296,14 @@ public:
       delete handler.w_notifier;
       delete handler.e_notifier;
     }
+    handlers.clear();
   }
+
+  ~PlugFrame() { cleanup(); }
 
   tresult resizeView(Steinberg::IPlugView* view, Steinberg::ViewRect* newSize) override
   {
-    wc.setSizeFromVst(*view, *newSize, w);
+    wc.setSizeFromVst(*view, *newSize, *w);
     return Steinberg::kResultOk;
   }
 };

--- a/src/plugins/score-plugin-vst3/Vst3/UI/PlugFrame.hpp
+++ b/src/plugins/score-plugin-vst3/Vst3/UI/PlugFrame.hpp
@@ -25,7 +25,7 @@ public:
   QDialog* w;
   WindowContainer wc;
   PlugFrame(QDialog& w, WindowContainer& wc)
-      : w{w}
+      : w{&w}
       , wc{wc}
   {
   }
@@ -55,7 +55,7 @@ public:
   QDialog* w;
   WindowContainer wc;
   PlugFrame(QDialog& w, WindowContainer& wc)
-      : w{w}
+      : w{&w}
       , wc{wc}
   {
   }

--- a/src/plugins/score-plugin-vst3/Vst3/UI/PlugFrame.hpp
+++ b/src/plugins/score-plugin-vst3/Vst3/UI/PlugFrame.hpp
@@ -22,7 +22,7 @@ public:
   Steinberg::uint32 addRef() override { return 1; }
   Steinberg::uint32 release() override { return 1; }
 
-  QDialog& w;
+  QDialog* w;
   WindowContainer wc;
   PlugFrame(QDialog& w, WindowContainer& wc)
       : w{w}
@@ -33,7 +33,7 @@ public:
   Steinberg::tresult
   resizeView(Steinberg::IPlugView* view, Steinberg::ViewRect* newSize) override
   {
-    wc.setSizeFromVst(*view, *newSize, w);
+    wc.setSizeFromVst(*view, *newSize, *w);
     return Steinberg::kResultOk;
   }
 };
@@ -52,7 +52,7 @@ public:
   Steinberg::uint32 addRef() override { return 1; }
   Steinberg::uint32 release() override { return 1; }
 
-  QDialog& w;
+  QDialog* w;
   WindowContainer wc;
   PlugFrame(QDialog& w, WindowContainer& wc)
       : w{w}
@@ -63,7 +63,7 @@ public:
   Steinberg::tresult
   resizeView(Steinberg::IPlugView* view, Steinberg::ViewRect* newSize) override
   {
-    wc.setSizeFromVst(*view, *newSize, w);
+    wc.setSizeFromVst(*view, *newSize, *w);
     return Steinberg::kResultOk;
   }
 };

--- a/src/plugins/score-plugin-vst3/Vst3/UI/Window.cpp
+++ b/src/plugins/score-plugin-vst3/Vst3/UI/Window.cpp
@@ -18,7 +18,14 @@ WindowContainer createVstWindowContainer(
   wc.qwindow = parentWindow.windowHandle();
   wc.container = nullptr;
 
-  wc.frame = new PlugFrame{parentWindow, wc};
+  if(!e.fx.plugFrame)
+    const_cast<Model&>(e).fx.plugFrame = new PlugFrame{parentWindow, wc};
+  else
+  {
+    e.fx.plugFrame->w = &parentWindow;
+    e.fx.plugFrame->wc = wc;
+  }
+  wc.frame = e.fx.plugFrame;
   view.setFrame(wc.frame);
   view.attached((void*)wc.qwindow->winId(), currentPlatform());
 

--- a/src/plugins/score-plugin-vst3/Vst3/UI/WindowCommon.cpp
+++ b/src/plugins/score-plugin-vst3/Vst3/UI/WindowCommon.cpp
@@ -22,9 +22,14 @@ Window::Window(const Model& e, const score::DocumentContext& ctx, QWidget* paren
 
 Window::~Window()
 {
+  if(m_model.fx.view)
+  {
+    m_model.fx.view->setFrame(nullptr);
+    m_model.fx.view->removed();
+  }
+
   if(container.frame)
   {
-    delete container.frame;
     container.frame = nullptr;
   }
 }
@@ -39,9 +44,9 @@ void Window::resizeEvent(QResizeEvent* event)
 void Window::closeEvent(QCloseEvent* event)
 {
   QPointer<Window> p(this);
+  m_model.fx.view->setFrame(nullptr);
   if(auto view = m_model.fx.view)
     view->removed();
-  delete container.frame;
   container.frame = nullptr;
 
   const_cast<QWidget*&>(m_model.externalUI) = nullptr;

